internal/jobobject/iocp.go

```diff
@@ -0,0 +1,83 @@
+package jobobject
+
+import (
+	"context"
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	ioInitOnce sync.Once
+	initIOErr  error
+	// Global iocp handle that will be re-used for every job object
+	ioCompletionPort windows.Handle
+	// Mapping of job handle to channel that that job object owns to be able to receive
+	// notifications.
+	jobMap = make(map[uintptr]chan interface{})
+)
+
+// AllProcessesExited is a type representing a notification that every process in a job has exited.
+type AllProcessesExited struct{}
+
+// ioCompletionProcessor processes completed async IOs forever
+func pollIOCP(ctx context.Context, iocpHandle windows.Handle) {
+	var (
+		overlapped uintptr
+		code       uint32
+		key        uintptr
+	)
+
+	for {
+		err := winapi.GetQueuedCompletionStatus(iocpHandle, &code, &key, (**windows.Overlapped)(unsafe.Pointer(&overlapped)), windows.INFINITE)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("failed during `GetQueuedCompletionStatus`")
```

Comment:
This log message is very generic and should be improved. I'd say something like `Failed to poll for job object notification`.